# Sentiment Analysis Part 2

_Natural Langauge Processing Nanodegree Program_

---



## Step 5: Switching gears - RNNs

We just saw how the task of sentiment analysis can be solved via a traditional machine learning approach: BoW + a nonlinear classifier. We now switch gears and use Recurrent Neural Networks, and in particular LSTMs, to perform sentiment analysis in Keras. Conveniently, Keras has a built-in [IMDb movie reviews dataset](https://keras.io/datasets/#imdb-movie-reviews-sentiment-classification) that we can use, with the same vocabulary size.

In [3]:
from keras.datasets import imdb  # import the built-in imdb dataset in Keras

# Set the vocabulary size
vocabulary_size = 5000

# Load in training and test data (note the difference in convention compared to scikit-learn)
(X_train, y_train), (X_test, y_test) = imdb.load_data(num_words=vocabulary_size)
print("Loaded dataset with {} training samples, {} test samples".format(len(X_train), len(X_test)))

Using TensorFlow backend.


17465344/17464789 [==============================] - 0s 0us/step
Loaded dataset with 25000 training samples, 25000 test samples


In [4]:
# Inspect a sample review and its label
print("--- Review ---")
print(X_train[7])
print("--- Label ---")
print(y_train[7])

--- Review ---
[1, 118, 20, 126, 14, 20, 3095, 61, 2, 43, 34, 4, 1147, 7, 2129, 21, 12, 16, 73, 290, 12, 13, 92, 2, 8, 81, 6, 2735, 7, 14, 321, 20, 43, 140, 67, 12, 48, 25, 28, 4, 580, 13, 104, 25, 80, 345, 119, 12, 42, 784, 12, 198, 4, 2432, 7, 6, 147, 1215, 20]
--- Label ---
1


Notice that the label is an integer (0 for negative, 1 for positive), and the review itself is stored as a sequence of integers. These are word IDs that have been preassigned to individual words. To map them back to the original words, you can use the dictionary returned by `imdb.get_word_index()`.

In [5]:
# Map word IDs back to words
word2id = imdb.get_word_index()
id2word = {i: word for word, i in word2id.items()}
print("--- Review (with words) ---")
print([id2word.get(i, " ") for i in X_train[7]])
print("--- Label ---")
print(y_train[7])

1646592/1641221 [==============================] - 0s 0us/step
--- Review (with words) ---
['the', 'where', 'on', 'your', 'as', 'on', 'catholic', 'only', 'and', 'out', 'who', 'of', 'rated', 'br', 'delivered', 'not', 'that', 'with', 'much', 'main', 'that', 'was', 'then', 'and', 'in', 'people', 'is', 'forgot', 'br', 'as', 'shot', 'on', 'out', 'through', 'can', 'that', 'what', 'have', 'one', 'of', 'lack', 'was', 'two', 'have', 'into', 'given', 'did', 'that', "it's", 'tale', 'that', 'series', 'of', 'australian', 'br', 'is', 'now', 'images', 'on']
--- Label ---
1


In [6]:
X_len=[len(v) for v in X_train]
print("Max length:{} and Min length: {}".format(max(X_len),min(X_len)))

Max length:2494 and Min length: 11


Unlike our Bag-of-Words approach, where we simply summarized the counts of each word in a document, this representation essentially retains the entire sequence of words (minus punctuation, stopwords, etc.). This is critical for RNNs to function. But it also means that now the features can be of different lengths!

#### Question: Variable length reviews

What is the maximum review length (in terms of number of words) in the training set? What is the minimum?

#### Answer:

maximum review length: 2494; 
minimu review length: 11


### TODO: Pad sequences

In order to feed this data into your RNN, all input documents must have the same length. Let's limit the maximum review length to `max_words` by truncating longer reviews and padding shorter reviews with a null value (0). You can accomplish this easily using the [`pad_sequences()`](https://keras.io/preprocessing/sequence/#pad_sequences) function in Keras. For now, set `max_words` to 500.

In [7]:
from keras.preprocessing import sequence

# Set the maximum number of words per document (for both training and testing)
max_words = 500

# TODO: Pad sequences in X_train and X_test
X_train =sequence.pad_sequences(X_train, maxlen=max_words)
X_test =sequence.pad_sequences(X_test, maxlen=max_words)

### TODO: Design an RNN model for sentiment analysis

Build your model architecture in the code cell below. We have imported some layers from Keras that you might need but feel free to use any other layers / transformations you like.

Remember that your input is a sequence of words (technically, integer word IDs) of maximum length = `max_words`, and your output is a binary sentiment label (0 or 1).

In [8]:
from keras.models import Sequential
from keras.layers import Embedding, LSTM, Dense, Dropout

# TODO: Design your model
model = Sequential()
input_dim =vocabulary_size
output_dim =32   #embedding_size
model.add(Embedding(input_dim,output_dim,input_length =max_words ))
model.add(LSTM(128))
model.add(Dense(1,activation='sigmoid'))
print(model.summary())

_________________________________________________________________
Layer (type)                 Output Shape              Param #   
embedding_1 (Embedding)      (None, 500, 32)           160000    
_________________________________________________________________
lstm_1 (LSTM)                (None, 128)               82432     
_________________________________________________________________
dense_1 (Dense)              (None, 1)                 129       
Total params: 242,561
Trainable params: 242,561
Non-trainable params: 0
_________________________________________________________________
None


#### Question: Architecture and parameters

Briefly describe your neural net architecture. How many model parameters does it have that need to be trained?

#### Answer:
mbedding_2 (Embedding)      (None, 500, 32)           160000    
_________________________________________________________________
lstm_2 (LSTM)                (None, 128)               82432     
_________________________________________________________________
Total params: 242,561


### TODO: Train and evaluate your model

Now you are ready to train your model. In Keras world, you first need to _compile_ your model by specifying the loss function and optimizer you want to use while training, as well as any evaluation metrics you'd like to measure. Specify the approprate parameters, including at least one metric `'accuracy'`.

In [9]:
# TODO: Compile your model, specifying a loss function, optimizer, and metrics
model.compile(loss='binary_crossentropy', optimizer='adam', metrics=['accuracy'])

Once compiled, you can kick off the training process. There are two important training parameters that you have to specify - **batch size** and **number of training epochs**, which together with your model architecture determine the total training time.

Training may take a while, so grab a cup of coffee, or better, go for a hike! If possible, consider using a GPU, as a single training run can take several hours on a CPU.

> **Tip**: You can split off a small portion of the training set to be used for validation during training. This will help monitor the training process and identify potential overfitting. You can supply a validation set to `model.fit()` using its `validation_data` parameter, or just specify `validation_split` - a fraction of the training data for Keras to set aside for this purpose (typically 5-10%). Validation metrics are evaluated once at the end of each epoch.

In [10]:
validation_size =5000 # 0.02* len(X_train)


In [11]:
# TODO: Specify training parameters: batch size and number of epochs

batch_size = 1024
num_epochs = 3

# TODO(optional): Reserve/specify some training data for validation (not to be used for training)
X_valid, y_valid = X_train[:validation_size], y_train[:validation_size]
X_train1, y_train1 =X_train[validation_size:], y_train[validation_size:]
# TODO: Train your model
model_trained =model.fit(X_train1,y_train1, 
          validation_data=(X_valid, y_valid),
          batch_size =batch_size, epochs =num_epochs, verbose=1)

Train on 20000 samples, validate on 5000 samples
Epoch 1/3
20000/20000 [==============================] - 1162s 58ms/step - loss: 0.6906 - acc: 0.5684 - val_loss: 0.6671 - val_acc: 0.5228
Epoch 2/3
20000/20000 [==============================] - 1175s 59ms/step - loss: 0.6689 - acc: 0.5773 - val_loss: 0.6667 - val_acc: 0.6338
Epoch 3/3
20000/20000 [==============================] - 1164s 58ms/step - loss: 0.6578 - acc: 0.6779 - val_loss: 0.6484 - val_acc: 0.7070


In [14]:
# Evaluate your model on the test set
scores = model.evaluate(X_test, y_test, verbose=0)  # returns loss and other metrics specified in model.compile()
print("Test accuracy:", scores[1])  # scores[1] should correspond to accuracy if you passed in metrics=['accuracy']

Test accuracy: 0.7034


In [13]:
#visual loss history
training_loss = model_trained.history['loss']
test_loss =model_trained.history['val_loss']
epoch_count =range(1, len(training_loss) + 1)

plt.plot(epoch_count, training_loss, 'r--')
plot.plot(epoch_count, test_loss, 'b-')
plt.legend(['Training Loss','Test_Loss'])
plt.xlabel('Epoch')
plt.ylabel('Loss')
plt.show()

NameError: name 'plt' is not defined

In [ ]:
import os
import pickle
cache_dir =os.path.join("cache","sentiment_analysis")
#os.makedirs(cache_dir, exit_ok=True)

In [ ]:
# Save your model, so that you can quickly load it in future (and perhaps resume training)
model_file = "rnn_model.h5"  # HDF5 file
model.save(os.path.join(cache_dir, model_file))

# Later you can load it using keras.models.load_model()
#from keras.models import load_model
#model = load_model(os.path.join(cache_dir, model_file))

Once you have trained your model, it's time to see how well it performs on unseen test data.

#### Question: Comparing RNNs and Traditional Methods

How well does your RNN model perform compared to the BoW + Gradient-Boosted Decision Trees?

#### Answer:

...

## Extensions

There are several ways in which you can build upon this notebook. Each comes with its set of challenges, but can be a rewarding experience.

- The first thing is to try and improve the accuracy of your model by experimenting with different architectures, layers and parameters. How good can you get without taking prohibitively long to train? How do you prevent overfitting?

- Then, you may want to deploy your model as a mobile app or web service. What do you need to do in order to package your model for such deployment? How would you accept a new review, convert it into a form suitable for your model, and perform the actual prediction? (Note that the same environment you used during training may not be available.)

- One simplification we made in this notebook is to limit the task to binary classification. The dataset actually includes a more fine-grained review rating that is indicated in each review's filename (which is of the form `<[id]_[rating].txt>` where `[id]` is a unique identifier and `[rating]` is on a scale of 1-10; note that neutral reviews > 4 or < 7 have been excluded). How would you modify the notebook to perform regression on the review ratings? In what situations is regression more useful than classification, and vice-versa?

Whatever direction you take, make sure to share your results and learnings with your peers, through blogs, discussions and participating in online competitions. This is also a great way to become more visible to potential employers!